### DISCUSSION TITLE
chap 6-2 질문 정리

### DISCUSSION THREAD

**Discussion by sihyung92** [JS-OS-study/hyokyung-and-students#20] [General]
process synchronzation 3,4

**Comment by 1chz** [comment]
# 느낀점

---

- 아직 뮤텍스, 세마포어, 모니터의 정확한 차이에 대해 완벽하게는 모르겠다
  - 스핀락: 락을 획득하지 못한 프로세스가 계속해서 락을 획득할 수 있는지 확인하게 되는 락.
  - 뮤텍스: 프로세스가 락을 획득한다면 임계영역에 들어가 작업하고, 락을 획득하지 못했다면 락이 해제 될 때까지 잠들게 하는 락으로 뮤텍스 락은 한 개만 존재한다.
  - 세마포어: 뮤텍스와 비슷하지만, 락의 개수가 한 개 이상일 수 있다.
- C언어를 배워 리눅스를 더 깊게 공부해봐야겠다는 생각이 들었다.

# 문제

---

- `Deadlock`이 발생하기 위한 필요 조건은 ______, ______, ______, ______ 이며, 이 중 단 하나라도 만족하지 않는다면 `Deadlock`은 발생하지 않습니다.

**Reply by sihyung92** [reply]
상호배제, 점유 및 대기, 비선점, 순환 대기

**Reply by inseo24** [reply]
상호배제, 점유 및 대기, 비선점, 순환 대기

**Reply by pullingoff** [reply]
상호 배제, 비선점, 점유와 대기, 원형 대기

**Comment by inseo24** [comment]
<h3>느낀 점 & 궁금한 점 </h3>

- 고전적인 synchronization 문제점을 알게 됐다. 
- 아직 모니터의 원리를 정확히 이해하지 못했다. 

<hr />
<h3>문제</h3>

Q. Monitor 는 세마포어와 달리 ____ 처리가 필요하지 않다.
A. _ _ _ _(영어로 4글자, 한글로 1글자)

**Reply by pullingoff** [reply]
lock

**Reply by 1chz** [reply]
lock

**Reply by sihyung92** [reply]
락

**Comment by pullingoff** [comment]
# 느낀 점
지난 차시에 내가 binary semaphore = mutex라고 했었다. 바이너리 세마포는 값이 0과 1 중 하나만 가능하기 때문에 '사용 가능한 자원이 있다 또는 없다'를 나타낸다. 뮤텍스 락과 유사하게 상호 배제의 기능을 가진다. 몇몇 시스템에선 아예 뮤텍스 락 없이 이진 세마포가 대신 상호 배제를 보장한다. 이 둘의 차이는 바로 락을 해제할 수 있는 프로세스이다. 뮤텍스의 경우 락을 건 프로세스 **자신**만이 락을 해제할 수 있는 반면 이진 세마포어의 경우 다른 프로세스가 락을 해제할 수 있다.

# 문제
1. bounded-buffer problem에서 소비자 프로세스가 데이터를 꺼내갈 때 데이터를 꺼낼 Full 버퍼에만 락을 건다. (O,X) (답: 6-3 강의 14분 30초)

2. dining philosophers 문제에서 데드락이 발생하는 경우는? (___ 철학자가 _____한 경우)

3. monitor에서 프로세스가 wait()할 때, 생산자 프로세스는 (꽉 찬/빈) 버퍼를 기다리는 큐에 줄을 서고 소비자 프로세스는 (꽉 찬/빈) 버퍼를 기다리는 큐에 줄을 선다.


# 질문
누군가에게 세마포어를 설명할 때, '여러 프로세스가 동시에 한 자원에 접근하는 걸 막고자 락을 걸 때, 프로세스가 사용 가능한 자원의 토탈 개수를 나타내는 변수'라고 설명하면 될까요?

**Reply by 1chz** [reply]
# 문제 
1. O
2. 모든 철학자가 동시에 한쪽 식기를 집어들었을 경우
3. 빈, 꽉 찬

# 질문
잠금 관련된 내용이 들어가야 할 것 같아요

**Reply by inseo24** [reply]
1. o
2. 동시에 모두 젓가락을 들 때 ?
3. 빈/ 꽉 찬
4. shared data의 개수(공유하는 데이터의 개수?)

**Reply by sihyung92** [reply]
1. O
2. 철학자들이 다 자기 왼쪽에 있는 젓가락 집은 경우 
3. 빈, 꽉 찬

**Reply by pullingoff** [reply]
counting semaphore: 순서 보장이 된다. 내부적으로 큐가 있어서 시그널 호출되면 하나씩 -- 하고 프로세스의 수행 순서를 보장해준다.
semaphore: 코드를 보면 자료구조로 구현돼있고, 보면 while문으로 락 을 걸 수 있게 되어있음. (locking을 할때 쓰이는 데이터 / 안정적으로 자원의 수를 체크할 수 있게 해주는 변수)

**Comment by sihyung92** [comment]
# 느낀 점
1. 추상적인 개념들이 코드적으로 어떻게 구현될 수 있는지 알게되어 좋았다.
2. 객체마다 달려있는 wait(), notify() 에 대해(무시하고 살았던) 알게 되었다!
3. 502 에바인거 같다.

# 문제
1. 자바의 monitor를 활용하려면 ______ 키워드를 사용하면 된다.

**Reply by 1chz** [reply]
synchronized

**Reply by inseo24** [reply]
synchronized

**Comment by pullingoff** [comment]
semaphore: 프로그래머의 실수가 영향을 끼칠 수 있다. (= 안정성 낮음)
monitor: 자바의 인스턴스 각각이 lock을 가진다. 자바 내부적으로 모니터를 가지고 있어 synchronized를 쓰면 각 인스턴스의 모니터를 사용할 수 있다.
